src/main/java/gov/nasa/pds/api/engineering/elasticsearch/Antlr4SearchListener.java

```diff
@@ -3,294 +3,195 @@
-		 
-		 for (SearchParser.QueryTermContext QTCtx : ctx.queryTerm()) {
-			 currentOrStatementQB.should(this.queryTermQBs.pollLast());
-		 }
-		 currentOrStatementQB.minimumShouldMatch(1);
-	 }
-	 
-	 
+
 	 @Override
-	 public void exitAndStatement(SearchParser.AndStatementContext ctx) {
-		 Antlr4SearchListener.log.debug("exit andStatement: " + ctx.getText());
-		 
-		 for (SearchParser.QueryTermContext QTCtx : ctx.queryTerm()) {
-			 this.andStatementQBs.getLast().must(this.queryTermQBs.pollLast());
-		 }
-		 
+	 public void enterOrStatement(SearchParser.OrStatementContext ctx)
+	 {
+		 this.conjunction = conjunctions.OR;
 	 }
 
 	 @Override
-	 public void enterComparison(SearchParser.ComparisonContext ctx) {
-	
-		this.comparisonQBs.add(QueryBuilders.boolQuery());
-		 
-		 // termQuery or rangeQuery
-		 this.elasticQuery = null;
-		
-		 // march eq, ne
-		 this.comparisonBoolean = true;
-		
-		 // range (gt, ge, lt, le)
-		 this.rangeOperator = null;
-		 this.includeBoundary = false;
-		 
+	 public void enterComparison(SearchParser.ComparisonContext ctx)
+	 {
+		 this.wildcard = ctx.VALUE() != null && (ctx.VALUE().getSymbol().getText().contains("*") || ctx.VALUE().getSymbol().getText().contains("?"));
  	 }
-	 
-	 
-	 
+
 	@Override
-	public void exitComparison(SearchParser.ComparisonContext ctx) {
-		Antlr4SearchListener.log.debug("exit comparison: " + ctx.getText());
+	public void exitComparison(SearchParser.ComparisonContext ctx)
+	{
+		final String left = ElasticSearchUtil.jsonPropertyToElasticProperty (ctx.FIELD(0).getSymbol().getText());
+		String right;
+		QueryBuilder comparator = null;
 
-		try {
-			
-			String comparisonLeftArg = 
-					ElasticSearchUtil.jsonPropertyToElasticProperty(ctx.FIELD().getSymbol().getText());
-			Object comparisonRightArg = null; 
-			
-			Method method;
-			QueryBuilder comparisonQB = null;
-			
-			// parse right arg of the comparison
-			if (ctx.STRINGVAL() != null) {
-				this.log.debug("parse string");
-				String rightArg = ctx.STRINGVAL().getSymbol().getText();
-				comparisonRightArg = rightArg.substring(1, rightArg.length() - 1);
-			
-			} else if (ctx.NUMBER() != null){
-				this.log.debug("parse number");
-				comparisonRightArg = Float.parseFloat(ctx.NUMBER().getSymbol().getText());
-					
-			}
-			
-		
-			// apply comparison (match or range)
-			if (this.elasticQuery == "termQuery") {					
-				comparisonQB = (QueryBuilder) QueryBuilders.termQuery(comparisonLeftArg, comparisonRightArg);
-			}
-			else if (this.elasticQuery == "rangeQuery") {
-				comparisonQB = (RangeQueryBuilder) QueryBuilders.rangeQuery(comparisonLeftArg);
-				Method methodRangeOperator = RangeQueryBuilder.class.getDeclaredMethod(this.rangeOperator, Object.class, boolean.class);
-				comparisonQB = (QueryBuilder) methodRangeOperator.invoke(comparisonQB, 
-						comparisonRightArg,
-						this.includeBoundary);
-			}
+		if (ctx.FIELD(1) != null) right = ctx.FIELD(1).getSymbol().getText();
```

Comment:
@al-niessner I am still confused by this part. I know you struggled to make the grammar work for us, but do we need 4 different cases for the second member of the comparison ? Could we name them more explicitelly if we need to keep them all ? Thanks

Reply:
It has to do with the way ANTLR works - unfortunately. If you look at Search.g4 while reading this it may help. There are filve tokens FIELD, NUMBER, OPERATOR, STRINGVAL, and VALUE. They are all similar as you know. If you take the simple statement 'a eq b' ANTLR sees a FIELD, eq as OPERATOR, and then b as FIELD> It is because it before STRINGVAL and VALUE in Search.g4. Hence, it needs to be on both left and right of the operator. Since '"' is not allowed in FIELD, ANTLR flows it to SRTINGVAL. Since '*' and '?' are not allowed in FIELD nor NUMBER nor STRINGVAL ANTLR makes it to VALUE. However, in 'a eq b', the first longest token 'b' can be assigned to is FIELD. Hence, it has to be on both the right and left side of the operator.

Reply:
Thanks @al-niessner , the order thing is what  I needed to get why we get FIELD as a second argument. Can you add that in the comments for the line 131 ?
The comments would be like, the second term of the comparison can be first tokenized as FIELD (string without quote) since FIELD is before VALUE (string with optional wildcard,  no quote) in the ANTLR4 grammar.

Thanks

Reply:
add requested comment